contracts/libs/Ibc.sol

```diff
@@ -172,17 +172,29 @@ library IbcUtils {
 
     // convert params to UniversalPacketBytes with optimal gas cost
 
-    function toUniversalPacketBytes(UniversalPacket memory data) internal pure returns (bytes memory packetBytes) {
-        packetBytes = abi.encode(data);
-    }
-
     // fromUniversalPacketBytes converts UniversalPacketDataBytes to UniversalPacketData, per how its packed into bytes
-    function fromUniversalPacketBytes(bytes memory data)
-        internal
+    function fromUniversalPacketBytes(bytes calldata data)
+        public
         pure
         returns (UniversalPacket memory universalPacketData)
     {
-        universalPacketData = abi.decode(data, (UniversalPacket));
+        bytes32 srcPortAddr;
+        uint256 mwBitmap;
+        bytes32 destPortAddr;
```

Comment:
Is it possible to avoid allocating these fields, instead directly unpack to UniversalPacket fields?

Reply:
I don't think so; I was trying this:

```
asembly {
            // Keep reusing 0x0 to move from calldata to return vars
            calldatacopy(0x0, data.offset, 32)
            universalPacketData.srcPortAddr := mload(0x0)
            calldatacopy(0x0, add(data.offset, 32), 32)
            universalPacketData.mwBitmap := mload(0x0)
            calldatacopy(0x0, add(data.offset, 64), 32)
            universalPacketData.destPortAddr := mload(0x0)
        }
```

But  the compiler doesn't like that, so that's why I went with just declaring the fields individually. Or were you thinking of some other way?

Reply:
Is it possible to do something like:

`universalPacketData = UniversalPacket(data[0:31], uint256(data[32:63], data[64:95], data[96:.length])`

?